rfcs/20200624-pluggable-device-for-tensorflow.md

```diff
@@ -0,0 +1,251 @@
+* Set 2: `PluggableDevicePlatform` related classes 
+   * class `PluggableDevicePlatform`: PluggableDevice-specific platform, its platform name is "PluggableDevice", it contains a C struct: SE_Platform* platform_ which is its internal implementation and as the C interface registered by device plugin.
+   * class `PluggableDeviceExecutor`: PluggableDevice-platform implementation of the platform-agnostic StreamExecutorInterface, it contains C structs: SE_StreamExecutor* executor_ and SE_Device* device_ whose member can be accessed in both TensorFlow proper and device plugins.
+   * class `PluggableDeviceStream`: wraps a StreamHandle in order to satisfy the platform-independent StreamInterface. It returns SE_Stream which is treated as an opaque type to TensorFlow,  whose structure is created by the device plugin.  
+   * class `PluggableDeviceTimer`: wraps an opaque handle: SE_Timer to satisfy the platform-independent TimerInterface.
+   * class `PluggableDeviceEvent`: wraps an opaque handle: SE_Event to satisfy the platform-independent EventInterface.
+
+**Tensorflow Plugin**
+
+Plugin authors need to provide those C functions implementation defined in StreamExecutor C API . 
+*  `SE_StreamExecutor` is defined as struct in the C API, both sides(TensorFlow proper and plugins) can access its members. Plugin creates the SE_StreamExecutor and registers its C API implementations to the SE_StreamExecutor.  
+```cpp
+   SE_StreamExecutor* create_stream_executor() {
+     SE_StreamExecutor* se_nfs = new SE_StreamExecutor();
+     se->memcpy_from_host = my_device_memory_from_host_function;
+     se->allocate = my_allocate_function;
+     …
+   }//Init device
+```
+* `SE_Device` is defined as struct in the C API, both sides(TensorFlow proper and plugins) can access its members. Plugin creates the SE_Device and fills its device opaque handle and device name to the SE_Device.
+```cpp
+  SE_Device* create_device(SE_Options* options, TF_Status* status) {
+    SE_Device* se = new SE_Device();
+    se->device_handle = get_my_device_handle();
+    ...
+    return se;
+  }
+```
+* `SE_Stream` is defined in plugin and treated as an opaque struct in TensorFlow proper. 
+```cpp
+  void create_stream(SE_Device* executor, SE_Stream* stream, TF_Status*) {
+    *stream = new SE_Stream_st();
+    (*stream)->stream_handle = create_my_stream_handle(executor);
+    ..
+  }
+```
+
+### PluggableDevice kernel registration
+
+This RFC shows an example of registering kernels for PluggableDevice. Kernel and op registration and implementation API is addressed in a separate [RFC](https://github.com/tensorflow/community/blob/master/rfcs/20190814-kernel-and-op-registration.md). 
+
+TensorFlow proper defines a new device_type named DEVICE_PLUGGABLE for PluggableDevice.This device_type is used for the kernel registration and dispatch. Plugin needs to register its kernel implementation with DEVICE_PLUGGABLE type.
```

Comment:
One concern I have here is that this doesn't quite allow the same TF process to have multiple pluggable devices, right?

If so we should call out this limitation early and disallow loading of more than one pluggable device shared object with a clear error message instead of ending up with conflicting kernel registrations.

Reply:
Agree,I will call out the limitation

Reply:
Currently we only have one pluggable device plugin support, but it can be extend to support multiple devices if StreamExecutor C API provides method to register the device type and those hard code part are clean up(such as [ProcessMemoryTypes](https://github.com/tensorflow/tensorflow/blob/d93971b09e699f646f773552d53cf3caad28bdea/tensorflow/core/common_runtime/memory_types.cc#L51) ), 
mulitple devices support may need more discussions 
&emsp;1. if two plugins are installed and both of them register themselves as 'GPU' device and same priority, device id should be unified.(gpu0-gpu3+ gpu0-3 -> gpu0->5)
&emsp;2. if two plugins are installed and both of them register themselves as "GPU" device and different priority, only one plugin will work.
&emsp;3. if two plugin are installed and register themselves with different name. this can be worked. user need to list device first then use its name to tf.device()